bokehjs/src/lib/models/widgets/file_input.ts

```diff
@@ -16,43 +16,59 @@ export class FileInputView extends WidgetView {
     if (this.dialogEl == null) {
       this.dialogEl = document.createElement('input')
       this.dialogEl.type = "file"
-      this.dialogEl.multiple = false
-      this.dialogEl.onchange = (e) => this.load_file(e)
+      this.dialogEl.multiple = this.model.multiple
+      this.dialogEl.onchange = async (e) => await this.load_files(e)
       this.el.appendChild(this.dialogEl)
     }
     if (this.model.accept != null && this.model.accept != '')
       this.dialogEl.accept = this.model.accept
+
     this.dialogEl.style.width = `{this.model.width}px`
     this.dialogEl.disabled = this.model.disabled
   }
 
-  load_file(e: any): void {
-    const reader = new FileReader()
-    this.model.filename = e.target.files[0].name
-    reader.onload = (e) => this.file(e)
-    reader.readAsDataURL(e.target.files[0])
+  async load_files(e: any): Promise<void> {
+    const files = e.target.files
+    const value: string[] = []
+    const filename: string[] = []
+    const mime_type: string[] = []
+    let i: number
+
+    for (i=0; i<files.length; i++){
+      filename.push(files[i].name)
+      const content = await this.readfile(files[i])
+      const file_arr = content.split(",")
+      value.push(file_arr[1])
+      mime_type.push(file_arr[0].split(":")[1].split(";")[0])
```

Comment:
This code needs to be refactored in a way that we have at least a vague understanding what it does (or alternatively, if you have to change it, let's say in a year, then you will be able to do it without any problems). So, starting from the beginning, `content` variable isn't really file's content, but a data URL. A better name would be e.g. `data_url`. Then all the further text splitting makes finally some sense to the reader, as we can infer that you are parsing said data URL. Then you extract URL's declaration and actual content. Again, `file_arr` isn't the best name, because it doesn't tell us anything about the contents of this variable, except for it being allegedly an array. We know, from data URL specification, that it contains exactly one `,`, thus we get a 2-tuple a result. Based on this knowledge, one could use `split(",", 2)` to improve performance on this line. Instead of introducing `file_arr`, it would be better to immediately unpack the result as `[declaration, content]`, resulting in code like this:

```ts
const data_url = await this.readfile(files[i])
const [declaration, content] = data_url.split(",", 2)
value.push(content)
mime_type.push(declaration.split(":")[1].split(";")[0])
```
Much better, though the last line is still a bit of a sore. That's just one approach. Alternatively you could use a regex to split the data URL in one go:

```ts
const data_url = await this.readfile(files[i])
const [, mime, , content] = data_url.split(/[:;,]/, 4)
value.push(content)
mime_type.push(mime)
```
Very readable, but performance may suffer (likely not if splitting limit is used, because the actual content isn't going to be scanned by `split()` method). It's up to you as the author to decide which is the best approach, considering criteria like correctness, readability (maintainability) and performance.

Reply:
Followed your advice and chose the later option.
The only change is I called "content" "data" instead, as it is the "data" part of the "data_url" in my view. 
Hope I found all the missing and excessive white spaces, too.
I tried to install and use tslint locally but had no success on that (so far).